### PR TITLE
Water Li

### PR DESCRIPTION
**Pull request by nerdyistrendy**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | spotlight method to find the work with the max votes
Describe how you approached testing that model method. What edge cases did you come up with? | edge cases including cases where no votes for a category/all work or when there is no work/user at all will the pages break. 
What are `session` and `flash`? What is the difference between them? | session keeps the data until you close the browser. Flash keeps the data for one request response cycle.
What was one thing that you gained more clarity on through this assignment? | one - many and many - many relationships
What is the Heroku URL of your deployed application? | https://bestmediaranker.herokuapp.com

# Assignment Submission: Media Ranker

### CODE REVIEW

File: app/models/vote.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class Vote < ApplicationRecord
+  belongs_to :user
+  belongs_to :work
+
+  validates :work_id, uniqueness: {scope: :user_id}
```
Comment (kaidamasaki) [inline app/models/vote.rb:5]: Nice job using the scope here!

File: app/views/homepages/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+<section class="spotlight">
+  <h2 class="spotlight__header">
+    <span class="spotlight__header--prefix">Media Spotlight:</span>
+    <% unless Work.spotlight.nil?  %>
+      <%= link_to Work.spotlight.title, work_path(Work.spotlight.id), class: "spotlight__link-to" %> by <%= Work.spotlight.creator %>
+      </h2>
+      <p class="spotlight__description">
+        <%= Work.spotlight.votes.count %> votes - <%= Work.spotlight.description %>
+      </p>
+    <% end  %>
```
Comment (kaidamasaki) [inline app/views/homepages/index.html.erb:11]: Because of where your `unless` is placed you wind up opening but never closing the `h2` if the spotlight is `nil`.

File: app/views/homepages/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        <ul class="list-group top-ten__list">
+          <% works.each do |work| %>
+            <li class="list-group-item">
+          <h4><%= link_to work.title, work_path(work.id) %>
+            <small class="top-ten__creator">by <%= work.creator %> </small>
+          </h4>
+              <p><%= work.votes.count%> votes</p>
+            </li>
+          <% end %>
+        </ul>
```
Comment (kaidamasaki) [inline app/views/homepages/index.html.erb:32]: While this is clever and works a partial would have been more appropriate here (and a little bit DRYer).

File: app/views/works/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,38 @@
+<main>
+
+<h2>List of Works</h2>
+
+<% hash = { "album" => "Albums", "book" => "Books", "movie" => "Movies"} %>
```
Comment (kaidamasaki) [inline app/views/works/index.html.erb:5]: This is another place where a partial would have been appropriate.

File: app/views/works/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    <%= link_to "Delete", work_path(@work), method: :delete, class: "btn btn-danger", data: {confirm: "Are you sure?"}%>
+  </section>
+
+  <section class="works-votes__container">
+    <h3 class="works-votes__header"> <%= @work.votes.count %> Votes for this <%= @work.category.capitalize %></h3>
+
+    <table class="table">
+      <thead>
+      <tr>
+        <th width = 50%>User</th>
```
Comment (kaidamasaki) [inline app/views/works/show.html.erb:20]: In the future you should avoid inline styling like this.  (The width should be set in the CSS.)

**Review by kaidamasaki — COMMENTED** [review]
Great job!  Your code was clean and clear and everything was perfectly functional.  😃 

# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | See inline notes.
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention | 
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention | 





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅
